src/ShellRouting/IShellRouter.cs

```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Jupyter.Core.Protocol;
+
+namespace Microsoft.Jupyter.Core
+{
+    public interface IShellHandler
+    {
+        public string MessageType { get; }
+        public void Handle(Message message);
+    }
+
+    public interface IShellRouter
```

Comment:
I like this one much better. We need to document how it works, though.

Reply:
Good point, I've added /// comments to help.